lib/model/compose.dart

```diff
@@ -0,0 +1,179 @@
+  //     > and no closing code fence has been found,
+  //     > the code block contains all of the lines after the opening code fence
+  //     > until the end of the containing block (or document).
+  //
+  // (B) Why not look for dangling closing fences (ones without an opening fence)?
+  //     Because technically there's no such thing:
+  //     they would be indistinguishable from dangling opening fences,
+  //     and parsers (and this function) will treat them that way.
+  final fenceLength = getUnusedOpeningBacktickFenceLength(content);
+
+  resultBuffer.write('`' * fenceLength);
+  if (infoString != null) {
+    resultBuffer.write(infoString);
+  }
+  resultBuffer.write('\n');
+  resultBuffer.write(content);
+  resultBuffer.write('\n');
+  resultBuffer.write('`' * fenceLength);
+  resultBuffer.write('\n');
+  return resultBuffer.toString();
+}
+
+const hashReplacements = {
+  "%": ".",
+  "(": ".28",
+  ")": ".29",
+  ".": ".2E",
+};
+
+// Corresponds to encodeHashComponent in Zulip web;
+// see web/shared/src/internal_url.js.
+String encodeHashComponent(String str) {
+  return Uri.encodeComponent(str)
+    .replaceAllMapped(RegExp(r'[%().]'), (Match m) => hashReplacements[m[0]!]!);
+}
+
+/// A URL to the given [Narrow], on `store`'s realm.
+///
+/// To include /near/{messageId} in the link, pass a non-null [nearMessageId].
+// Why take [nearMessageId] in a param, instead of looking for it in [narrow]?
+//
+// A reasonable question: the "near" part of a near link (e.g., for
+// quote-and-reply) takes the same form as other operator/operand pairs that we
+// represent with [ApiNarrowElement]s, like "/stream/48-mobile".
+//
+// But unlike those other elements, we choose not to give the "near" element
+// an [ApiNarrowElement] representation, because it doesn't have quite that role:
+// it says where to look in a list of messages, but it doesn't filter the list down.
+// In fact, from a brief look at server code, it seems to be *ignored*
+// if you include it in the `narrow` param in get-messages requests.
+// When you want to point the server to a location in a message list, you
+// you do so by passing the `anchor` param.
+Uri narrowLink(PerAccountStore store, Narrow narrow, {int? nearMessageId}) {
+  final apiNarrow = narrow.apiEncode();
+  final resultBuffer = StringBuffer('#narrow');
+  for (final element in apiNarrow) {
+    resultBuffer.write('/');
+    if (element.negated) {
+      resultBuffer.write('-');
+    }
+    // TODO(server-7) remove special-casing
+    if (element is ApiNarrowDm) {
+      final supportsOperatorDm = store.connection.zulipFeatureLevel! >= 177; // TODO(server-7)
+      final resolved = element.resolve(legacy: !supportsOperatorDm);
+      final operator = resolved.operator;
+      final operand = resolved.operand;
+      final operandSuffix = operand.length >= 3
+        ? 'group'
+        : (supportsOperatorDm ? 'dm' : 'pm'); // TODO(?) CZO actually gives a slugified full_name here…??
```

Comment:
Huh. Looks like there's two implementations of generating a URL for a DM narrow: `pm_perma_link` and `pm_with_url`, both in `web/src/people.js`. The former does what this code does, and the latter does what you're seeing:
```js
    if (user_ids.length > 1) {
        suffix = "group";
    } else {
        const person = get_by_user_id(user_ids[0]);
        if (person && person.full_name) {
            suffix = person.full_name.replaceAll(/[ "%/<>`\p{C}]+/gu, "-");
```

I'm fine with just "dm". I think the fancier thing is most helpful for putting a URL in the location bar, which the user may see all the time when they're looking at that conversation.

Reply:
Ohh, I think I see why there's both. The thing that's happening is that `pm_with_url` includes only the *other* people in the thread, while `pm_perma_link` includes all including self. Which is necessary for a context like quote-and-reply, where the point of the link is likely for other people to be able to use it.

But then once it's listing all recipients and not just others, that means that in a 1:1 thread it's listing both people. So to use names, you'd have to write both names. Which we don't do even in the location bar for 3-person group DM threads, and so don't do for these permalinks either.

(A shame web is so thin on jsdoc or other comments.)

Anyway, so that just confirms the conclusion: stick with plain `dm`/`pm`.